#7560: Fix FD noc api to set multicast_path_reserve when mcasting

tt-aho opened the pull request:
### Ticket
https://github.com/tenstorrent/tt-metal/issues/7560

### Problem description
Stable diffusion was still seeing hangs after di/dt fixes. Was narrowed down to a commit that changed the FD eth cores used for prefetch/dispatch, which affected timing. The real latent bug was that multicast_path_reserve was not being set with the new noc apis used in some FD commands, leading to potential NOC deadlock.

### What's changed
Always set multicast_path_reserve for now for all mcast transactions with the new apis.

### Checklist
- [x] Post commit CI passes
- [x] Model regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

tt-aho committed:
#7560: Fix FD noc api to set multicast_path_reserve when mcasting

pgkeller reviewed (approved):
wow.  damn.  5 characters

pgkeller commented on tt_metal/impl/dispatch/kernels/cq_common.hpp:
```diff
@@ -124,7 +124,7 @@ void cq_noc_async_write_init_state(uint32_t src_addr, uint64_t dst_addr, uint32_
     constexpr bool posted = false;
     constexpr bool linked = false;
-    constexpr uint32_t vc = mcast ? NOC_MULTICAST_WRITE_VC : NOC_UNICAST_WRITE_VC;
+    constexpr uint32_t vc = mcast ? NOC_DISPATCH_MULTICAST_WRITE_VC : NOC_UNICAST_WRITE_VC;
```
I would make this a subsequent commit for easy regression testing

tt-aho replied:
Done

pgkeller reviewed (changes requested):
1 minor change, 1 question

tt-aho committed:
#0: Convert write_linear to use new dispatch NOC apis, in preparation to switch to using NOC_DISPATCH_MULTICAST_WRITE_VC for all dispatch mcasts
Remove some minor riscv overhead since we don't need to create full noc addrs if we're not reprogramming the noc coords

tt-aho committed:
#0: Switch all dispatch mcasts to use NOC_DISPATCH_MULTICAST_WRITE_VC

tt-aho commented on tt_metal/impl/dispatch/kernels/cq_dispatch.cpp:
```diff
@@ -381,27 +381,33 @@ void process_write_linear(uint32_t num_mcast_dests) {
                 // since they were all accounted for.
-                block_noc_writes_to_clear[rd_block_idx] -= (orphan_size == 0);
+                writes -= (orphan_size == 0);
+                mcasts -= (orphan_size == 0) * num_mcast_dests;
```
These should get removed I think once we move to writing up to fence instead of 4k transactions. Currently I just copied the impl from our other fns. Perf ci is passing though havent looked at actual perf differences yet.

tt-asaigal replied:
Removed on this branch: https://github.com/tenstorrent/tt-metal/tree/asaigal/remove_orphan_writes.
